### PR TITLE
feat: introduce new monetary int

### PR DESCRIPTION
**Pull request by altitude**

### CODE REVIEW

File: Taskfile.yaml [hunk 1]
```diff
@@ -42,7 +42,7 @@ tasks:
     deps: [postgres]
     cmds:
       - >
-        go test {{.TAGS}} -v {{.FAILFAST}} -coverpkg {{.PKG}} -coverprofile coverage.out -covermode atomic
```
Comment (antoinegelloz) [inline Taskfile.yaml:45]: I like to have the verbose mode in local testing, why do you want to remove it?
If you don't want logs, I prefer that you remove the verbose mode in the `tests` rule above, or create a `VERBOSE: "-v"` variable that can be overridden with `tasks VERBOSE=""` :)

File: cmd/script_exec.go [hunk 2]
```diff
@@ -82,7 +82,13 @@ func NewScriptExec() *cobra.Command {
 			fmt.Printf("ID: %d\r\n", result.Transaction.ID)
 			fmt.Println("Postings:")
 			for _, p := range result.Transaction.Postings {
-				fmt.Printf("\t Source: %s, Destination: %s, Amount: %d, Asset: %s\r\n", p.Source, p.Destination, p.Amount, p.Asset)
+				fmt.Printf(
+					"\t Source: %s, Destination: %s, Amount: %s, Asset: %s\r\n",
+					p.Source,
+					p.Destination,
+					p.Amount.String(),
```
Comment (antoinegelloz) [inline cmd/script_exec.go]: No need to call the `.String()` method, fmt will call it automatically thanks to the fmt.Stringer interface :)

File: pkg/core/monetary.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+func (a *MonetaryInt) Gte(b *MonetaryInt) bool {
+	return (*big.Int)(a).Cmp((*big.Int)(b)) >= 0
+}
+
+func (a *MonetaryInt) Lt(b *MonetaryInt) bool {
+	return (*big.Int)(a).Cmp((*big.Int)(b)) < 0
+}
+
+func (a *MonetaryInt) Ltz() bool {
```
Comment (antoinegelloz) [inline pkg/core/monetary.go:58]:
```suggestion
func (a *MonetaryInt) IsNeg() bool {
```
Wouldn't be more clear?

File: pkg/core/monetary.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	return (*big.Int)(a).UnmarshalText(b)
+}
+
+func NewMonetaryInt(i int64) *MonetaryInt {
+	return (*MonetaryInt)(big.NewInt(i))
+}
+
+func ParseMonetaryInt(s string) (*MonetaryInt, error) {
+	i, ok := big.NewInt(0).SetString(s, 10)
+
```
Comment (antoinegelloz) [inline pkg/core/monetary.go]: useless line break

File: pkg/ledger/executor_test.go [hunk 5]
```diff
@@ -10,12 +10,12 @@ import (
...
 
-func assertBalance(t *testing.T, l *Ledger, account, asset string, amount int64) {
+func assertBalance(t *testing.T, l *Ledger, account, asset string, amount *core.MonetaryInt) {
 	user, err := l.GetAccount(context.Background(), account)
 	require.NoError(t, err)
 
 	b := user.Balances[asset]
-	assert.Equalf(t, amount, b,
+	assert.Equalf(t, amount.String(), b.String(),
 		"wrong %v balance for account %v, expected: %d got: %d",
```
Comment (antoinegelloz) [inline pkg/ledger/executor_test.go]:
```suggestion
		"wrong %v balance for account %v, expected: %s got: %s",
```

File: pkg/ledger/process_test.go [hunk 6]
```diff
@@ -258,18 +278,19 @@ func TestLedger_processTx(t *testing.T) {
...
 						PostCommitVolumes: core.AccountsAssetsVolumes{
 							"toto":  core.AssetsVolumes{"EUR": core.Volumes{Input: worldTotoEUR, Output: totoAliceEUR}},
-							"alice": core.AssetsVolumes{"EUR": core.Volumes{Input: worldAliceEUR + totoAliceEUR, Output: 0}},
+							"alice": core.AssetsVolumes{"EUR": core.Volumes{Input: worldAliceEUR.Add(totoAliceEUR), Output: core.NewMonetaryInt(0)}},
 						},
 					},
 				}
 
+				// b, _ := json.Marshal(expectedTxs)
+				// a, _ := json.Marshal(res.GeneratedTransactions)
```
Comment (antoinegelloz) [inline pkg/ledger/process_test.go]:
```suggestion
```
Oopsy ;)

File: pkg/ledger/volume_agg.go [hunk 7]
```diff
@@ -19,10 +19,21 @@ func (tva *transactionVolumeAggregator) postCommitVolumes() core.AccountsAssetsV
 }
 
 func (tva *transactionVolumeAggregator) preCommitVolumes() core.AccountsAssetsVolumes {
+	// for account, _ := range tva.preVolumes {
+	// 	for asset, _ := range tva.preVolumes[account] {
+	// 		if tva.preVolumes[account][asset].Input == nil {
+	// 			tva.preVolumes[account][asset].Input = core.NewMonetaryInt(0)
+	// 		}
+	// 	}
+	// }
```
Comment (antoinegelloz) [inline pkg/ledger/volume_agg.go]:
```suggestion
```

File: pkg/storage/sqlstorage/migrates/9-add-pre-post-volumes/any_test.go [hunk 8]
```diff
@@ -141,48 +161,61 @@ var testCases = []testCase{
...
 				"USD": {
-					Input: 200,
+					Input:  core.NewMonetaryInt(200),
+					Output: core.NewMonetaryInt(0),
 				},
-				"USDT": {},
+				// "USDT": {
+				// 	Input:  core.NewMonetaryInt(0),
+				// 	Output: core.NewMonetaryInt(0),
+				// },
```
Comment (antoinegelloz) [inline pkg/storage/sqlstorage/migrates/9-add-pre-post-volumes/any_test.go]:
```suggestion
```

File: pkg/storage/sqlstorage/migrates/9-add-pre-post-volumes/any_test.go [hunk 9]
```diff
@@ -141,48 +161,61 @@ var testCases = []testCase{
...
-				},
-				"USDT": {
-					Output: 90,
+					Input:  core.NewMonetaryInt(200),
+					Output: core.NewMonetaryInt(100),
 				},
+				// "USDT": {
+				// 	Input:  core.NewMonetaryInt(0),
+				// 	Output: core.NewMonetaryInt(0),
+				// },
```
Comment (antoinegelloz) [inline pkg/storage/sqlstorage/migrates/9-add-pre-post-volumes/any_test.go]:
```suggestion
```

**Review by antoinegelloz — CHANGES_REQUESTED** [review]
Looks good to me!
Just think about removing the testlog.txt file :)

**Review by flemzord — DISMISSED** [review]

File: Taskfile.yaml (hunk 1 above)
Comment (gfyrag) [reply Taskfile.yaml:45]: Pretty agree with @altitude the verbose logs complicate things most of the time.

File: pkg/ledger/executor_test.go (hunk 5 above)
Comment (gfyrag) [reply pkg/ledger/executor_test.go:19]: @antoinegelloz Why this modification?

**Review by gfyrag — APPROVED** [review]

File: pkg/core/monetary.go (hunk 3 above)
Comment (gfyrag) [reply pkg/core/monetary.go:58]: Ltz is a common term. I think it is not a problem.

File: Taskfile.yaml (hunk 1 above)
Comment (antoinegelloz) [reply Taskfile.yaml:45]: Ok then, can we change the committed version to `Taskfile.dist.yml`, so I can override it with my own locally?
https://taskfile.dev/usage/#supported-file-names

File: pkg/core/monetary.go (hunk 3 above)
Comment (antoinegelloz) [reply pkg/core/monetary.go:58]: Ok!

File: Taskfile.yaml (hunk 1 above)
Comment (gfyrag) [reply Taskfile.yaml:45]: Add variable VERBOSE to the Taskfile.
Give : 
```task tests VERBOSE=1```

**Review by gfyrag — DISMISSED** [review]

**Review by antoinegelloz — DISMISSED** [review]

**Review by jdupas22 — DISMISSED** [review]

**Review by antoinegelloz — APPROVED** [review]

**Review by jdupas22 — APPROVED** [review]

**Pull request merged**